Preflight Checks doc (#2730)
aws/eks-anywhere: designs/preflight-checks.md

                └── createcluster_test.go

```

Primary validation components called by validations/createcluster include the following:

* General `create` validations under `pkg/validations/createvalidations` ([link](https://github.com/aws/eks-anywhere/blob/b4a4eb84c03091d1a646c1e49b9760b8b63961d3/pkg/validations/createvalidations/preflightvalidations.go))
    * `validate` will continue to run the general `create` validations in `validations/createvalidations` as part of the preflight checks to ensure a single source of truth for maintenance and consistency in user experience

* Provider specific validations under pkg/providers ([link](https://github.com/aws/eks-anywhere/blob/621a05272404b1d17c15e040007b4a14c9d0499b/pkg/providers/vsphere/vsphere.go#L372))

* Addonmanager/addonclients validation ([link](https://github.com/aws/eks-anywhere/blob/b4a4eb84c03091d1a646c1e49b9760b8b63961d3/pkg/addonmanager/addonclients/fluxaddonclient.go#L289))
    * fluxaddonclient
* Additional individual validation functions in validations/createcluster called directly to replicate `create cluster` validation tasks

#### Usage:

`exp` - subcommand, introduce all subcommands as experimental

> `eks-a exp validate create cluster [flags]`

Reviewer: ```suggestion
`eksctl anywhere exp validate create cluster [flags]`
```

# Preflight Validate Design Document

* * *

## Introduction

**Problem:** Users do not have a method to validate configuration files to be used with EKS Anywhere without attempting to create a cluster. 

With more complex provider configurations, users may need to review and modify their configuration multiple times before arriving at a valid configuration for cluster creation. Running the create cluster command for validation commits the user to attempting to create a cluster and generates artifacts that need to be removed.

### Tenets

* ****Simple:**** simple to use, simple to understand, simple to maintain
* **Informative**: output should lead user to take action and guide next steps 

> * **Non-invasive:** minimal permanent changes should be left as a result of this comman**d**

Reviewer 2: ```suggestion
* **Non-invasive:** minimal permanent changes should be left as a result of this command
```

* **Intuitive:** validation should be accomplished following the same patterns users have come to expect with EKS-A CLI

### Goals and objectives

As a Kubernetes administrator I want to:

* Validate EKS Anywhere configurations without running `create cluster`
* Validate required resources are available prior to cluster creation 
* Receive informative feedback about configuration and resource readiness

### Statement of scope

**In scope**

> * Preflight validation checks can be run independently of `create cluster`

Reviewer 2: ```suggestion
```

This is a goal :)

Ideally, the dependency generation process should be separate from the their validation to enable this architecture and allow the generation and validations to be called independently and or together as needed.

#### Sequence Diagram

![preflight checks sequence diagram](images/preflight-checks-sequence.png)
#### Validation Location

The `validate` command includes validations from the following locations originally based off the `create cluster` command validations. This does not include all code dependencies for the `validate` command. 

```

`.`
`└──`` eks``-``anywhere`
`    ``├──`` cmd``/``eks``-``a``/``cmd``
    │ ├── experiment.go 
    │ ├── validate.go 
    │ └── validatecreatecluster.go
``    ``└──`` pkg`
`        ``├──`` addonmanager``/``addonclients`
`        ``│``   ``└──`` fluxaddonclient``.``go`
`        ``├──`` api``/``v1alpha1`
`        ``│``   ``└──`` cluster``.``go`
`        ``├──`` providers`
`        ``│``   ``├──`` cloudstack`
`        ``│``   ``├──`` docker`
`        ``│``   ``├──`` snow`
`        ``│``   ``├──`` tinkerbell`
`        ``│``   ``└──`` vsphere`
`        ``└──`` validations`

> `            ``├── validatevalidations`

Reviewer 3: in this package new? The name is a bit weird

Reviewer 3: can they just be in `validations`?
Or maybe depending on what they are validating they can go in there own packages? Like if it's a docker validation, in a docker package, etc.

@chrisdoherty4 is better than be at naming packages

Reviewer 2: If its a set of validation helpers `validation` is fine. Initially a `validation.go` is fine and if we start finding groups of algorithms like 'network validation' then we can isolate that in its own source file down the road.

I wouldn't follow the `createvalidations` thing - that's rather odd.

Author: targeted validation helpers under `validation` folder but ended up making something pretty similar to the `createvalidations` flow - which as Chris mentioned may not be what we're shooting for. Will follow up to see what makes sense to move/restructure.

Author: Updated to move away from the `createvalidations` flow, currently this just consists of a `createcluster` pkg in validations that runs the associated validations in sequence.
